app/Repositories/BuildRepository.php

```diff
@@ -29,40 +30,68 @@ class BuildRepository
      */
     public function create(Application $application, Platform $platform, array $attributes)
     {
-        /** @var UploadedFile $buildFile */
-        $buildFile = Arr::get($attributes, 'file');
-        $availableFrom = Carbon::parse(Arr::get($attributes, 'available_from'));
+        return DB::transaction(function () use ($attributes, $platform, $application) {
 
-        return DB::transaction(function () use ($availableFrom, $attributes, $platform, $application, $buildFile) {
-            $buildPath = $buildFile->storeAs(
-                $application->slug,
-                $application->slug.'-'.$platform->getId().'-'.Arr::get(
-                    $attributes,
-                    'version'
-                ).'.'.$buildFile->getClientOriginalExtension(),
-                ['disk' => config('filesystems.cloud')]
-            );
+            /** @var UploadedFile $artifact */
+            $artifact = Arr::get($attributes, 'file');
+            $availableFrom = Carbon::parse(Arr::get($attributes, 'available_from'));
+
+            $path = $this->storeArtifact($artifact, $application, $platform, Arr::get(
+                $attributes,
+                'version'
+            ));
 
             /** @var Build $build */
             $build = $application->builds()->create([
                 'platform' => $platform->getId(),
                 'version' => Arr::get($attributes, 'version'),
-                'file' => $buildPath,
+                'file' => $path,
                 'forced' => Arr::get($attributes, 'forced', 'false') === 'true',
                 'available_from' => $availableFrom->toDateTimeString(),
             ]);
 
-            foreach (Arr::get($attributes, 'changelogs', []) as $locale => $content) {
-                $build->changelogs()->create([
-                    'locale' => $locale,
-                    'content' => $content
-                ]);
-            }
+            $this->saveChangelogs($build, Arr::get($attributes, 'changelogs', []));
 
             return $build;
         });
     }
 
+    /**
+     * @param  Build  $build
+     * @param  array  $attributes
+     * @return mixed
+     */
+    public function update(Build $build, array $attributes)
+    {
+        return DB::transaction(function () use ($build, $attributes) {
+            $build->update($attributes);
+
+            $changelogs = Arr::get($attributes, 'changelogs');
+
+            // storing updated changelogs
+            if ($changelogs !== null) {
+                $build->changelogs()->delete();
+                $this->saveChangelogs($build, $changelogs);
+            }
+
+            /** @var UploadedFile $artifact */
+            $artifact = Arr::get($attributes, 'file');
+
+            // if the new file is specified, store it
+            if ($artifact !== null) {
+                Storage::cloud()->delete($build->file);
```

Comment:
What about renaming it instead of deleting it so that in the future, we could easily create a rollback feature? What do you think?

Reply:
:thinking: if we wanted to keep a history I think we have to save this information to database in some way, currently once changed the path of the build file you lose the reference to what was before, leaving the previous build file "orphan". I think maybe we should do this in a dedicated task.

Reply:
Yep, makes sense